libvcx/src/api_lib/api_c/issuer_credential.rs

```diff
@@ -226,6 +274,94 @@ pub extern fn vcx_issuer_get_credential_offer_msg(command_handle: CommandHandle,
     error::SUCCESS.code_num
 }
 
+
+#[no_mangle]
+pub extern fn vcx_issuer_build_credential_offer_msg(command_handle: CommandHandle,
+                                                    credential_handle: u32,
+                                                    cred_def_handle: u32,
+                                                    credential_data: *const c_char,
+                                                    comment: *const c_char,
+                                                    cb: Option<extern fn(xcommand_handle: CommandHandle, err: u32, msg: *const c_char)>) -> u32 {
+    info!("vcx_issuer_build_credential_offer_msg >>>");
+
+    check_useful_c_callback!(cb, VcxErrorKind::InvalidOption);
+    check_useful_c_str!(credential_data, VcxErrorKind::InvalidOption);
+    let comment = if !comment.is_null() {
+        check_useful_opt_c_str!(comment, VcxErrorKind::InvalidOption);
+        comment.to_owned()
+    } else {
+        None
+    };
+
+    let source_id = issuer_credential::get_source_id(credential_handle).unwrap_or_default();
+    trace!("vcx_issuer_build_credential_offer_msg(command_handle: {}, credential_handle: {}) source_id: {}",
+           command_handle, credential_handle, source_id);
+
+    if !issuer_credential::is_valid_handle(credential_handle) {
+        return VcxError::from(VcxErrorKind::InvalidIssuerCredentialHandle).into();
+    }
+
+    execute(move || {
+        match issuer_credential::build_credential_offer_msg(credential_handle, cred_def_handle, credential_data, comment) {
+            Ok(offer_msg) => {
+                let offer_msg = json!(offer_msg).to_string();
+                let offer_msg = CStringUtils::string_to_cstring(offer_msg);
+                trace!("vcx_issuer_build_credential_offer_msg_cb(command_handle: {}, credential_handle: {}, rc: {}) source_id: {}",
+                       command_handle, credential_handle, error::SUCCESS.message, source_id);
+                cb(command_handle, error::SUCCESS.code_num, offer_msg.as_ptr());
+            }
+            Err(x) => {
+                warn!("vcx_issuer_build_credential_offer_msg_cb(command_handle: {}, credential_handle: {}, rc: {}) source_id: {})",
+                      command_handle, credential_handle, x, source_id);
+                cb(command_handle, x.into(), ptr::null_mut());
+            }
+        };
+
+        Ok(())
+    });
+
+    error::SUCCESS.code_num
+}
+
+#[no_mangle]
+pub extern fn vcx_mark_credential_offer_msg_sent(command_handle: CommandHandle,
```

Comment:
I think exposing a function which allows to move to `OfferSent` state without assuring the offer gets actually sent is asking for trouble (e.g. what if the transmission of the OOB message fails?). Another option (which is preferrable IMO) is to allow receiving credential request directly from both `OfferSet` and `OfferSent` states; this would also make this function redundant.

Reply:
Well, there's no way assure that it was sent, it's up to API consumer to assure that the message was sent through some communication channel. if we were to accept `credential request` in both state, we might as well just reduce the states to 1 state `OfferSet`. But with `vcx_mark_credential_offer_msg_sent` + `OfferSent`, the developer is forced to explicitly signal that the message was sent and to expect `Credential Request`. I find it safer than implicitly accept a credential request as soon as credential request is set.